### PR TITLE
[SPARK-8088] don't attempt to lower number of executors by 0

### PR DESCRIPTION
**Pull request by ryan-williams**

### CODE REVIEW

**Comment by andrewor14** [comment]
lgtm we have a similar check in the add case (L320)

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 1]
```diff
@@ -266,10 +266,12 @@ private[spark] class ExecutorAllocationManager(
       // executors and inform the cluster manager to cancel the extra pending requests
       val oldNumExecutorsTarget = numExecutorsTarget
       numExecutorsTarget = math.max(maxNeeded, minNumExecutors)
-      client.requestTotalExecutors(numExecutorsTarget)
-      numExecutorsToAdd = 1
-      logInfo(s"Lowering target number of executors to $numExecutorsTarget because " +
-        s"not all requests are actually needed (previously $oldNumExecutorsTarget)")
+      if (numExecutorsTarget < oldNumExecutorsTarget) {
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: We still want to set `numExecutorsToAdd` to 0, right?

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: hm yeah we should do that outside

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: @ryan-williams then can you also add a comment above this saying something like
```
// If the new target has not changed, avoid sending a message to the cluster manager
```

Comment (ryan-williams) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: @sryza I thought about that, but wasn't sure it was necessary.

You're worried about the case where, after executing L268, we have `maxNeeded < oldNumExecutorsTarget == numExecutorsTarget == minNumExecutors`, i.e. we need fewer than the minimum, we were already at the minimum, and will continue to be at the minimum.

If we're already at the minimum, we must have actually decreased to it at some point (and therefore set `numExecutorsToAdd = 1`), or we just started, in which case `numExecutorsToAdd == 1`).

@andrewor14 sure I'll add the comment

Comment (ryan-williams) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: basically, if the contract is "we start `numExecutorsToAdd` at 1, and every time we decrease we reset it to 1", then I think we get everything we want, no?

obvs it's a nit to pick, just want to be sure we're all agreed about the semantics and more importantly that I understand the code I'm modifying :)

**Comment by ryan-williams** [comment]
[previous test failure](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/34121/console) seems like Jenkins had some kind of hiccup fwiw

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala (hunk 1 above)
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: @ryan-williams that may be so, but it's much simpler to reason about if we just set `numExecutorsToAdd = 1` every time our target exceeds the number needed. I would just do it outside for that reason if it doesn't change any functionality.

Comment (ryan-williams) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: @andrewor14 makes sense, done

**Comment by sryza** [comment]
LGTM pending jenkins

**Comment by andrewor14** [comment]
master 1.4 thanks

**Pull request closed without merging**